### PR TITLE
docs: tutorial monitoring bias and ethics

### LINKED ISSUE #4447: [DOCS] Monitoring and curating LLMs for ethics and bias using RLHF
## Which page or section is this issue related to?

https://www.giskard.ai/

"Monitoring and curating LLMs for ethics and bias using RLHF" I looked at your LLM monitoring alpha solution an to me it makes sense to forwards the triggered sentences and the alternatives to Argilla to rank the, reward them or provide a correction. Eventually we can use the ArgillaTrainer to finetune them using TRL using RLHF

### PR DESCRIPTION
**Pull request by sdiazlor**
# Description

Include a new tutorial on monitoring bias and ethics using Giskard outputs

Closes #4447 

**Type of change**

(Remember to title the PR according to the type of change)

- [ ] Documentation update

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes.)

- [x] `sphinx-autobuild` (read [Developer Documentation](https://docs.argilla.io/en/latest/community/developer_docs.html#building-the-documentation) for more details)

**Checklist**

- [ ] I added relevant documentation
- [ ] I followed the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See https://keepachangelog.com/)

### CODE REVIEW

**Comment by sdiazlor** [comment]
@davidberenstein1957 Hi, David, this is the draft version, let me know what can be improved :)

**Comment by davidberenstein1957** [comment]
Hi @sdiazlor , looking great already. 

maybe something like this would be better, also because we don't do full RLHF?
👀 Monitoring and SFT LLMs for ethics and bias using Giskard

Could you add an output example of Giskard and Langchain? This helps to understand better what is happening.

Do you think you would be able to replace the reward model part section on Phi2, we might be able to outline code to fine-tune it using DPO?
- https://huggingface.co/blog/dpo-trl
- https://huggingface.co/microsoft/phi-2/discussions/19
- https://github.com/mit1280/fined-tuning/blob/main/phi_2_fine_tune_using_PEFT%2Binference.ipynb

If this proves too complex, we can also leave the last part for a specific tutorial on DPO/PPO, and focus on the monitoring part for this one.

File: docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb [hunk 1]
```diff
@@ -0,0 +1,628 @@
...
+   "metadata": {},
+   "source": [
+    "# 👀 Monitoring and curating LLMs for ethics and bias using RLHF"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In this tutorial, we explore a method to address `bias` in language models by enhancing the input data using `Reinforcement Learning with Human Feedback`. This approach aligns with ethical considerations by actively involving human judgment to guide the learning process, ensuring a more balanced and fair representation in the model's outputs.\n",
```
Comment (ignacioct) [inline docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb]: I wouldn't write bias as code, maybe bold if you wanna highlight it.

File: docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb [hunk 2]
```diff
@@ -0,0 +1,628 @@
...
+    "\n",
+    "The steps are as follows:\n",
+    "\n",
+    "1. Test our LLM using [Giskard](https://www.giskard.ai/) and analyze the results\n",
+    "2. Create an Argilla `Feedback Dataset` according to the outputs of Giskard\n",
```
Comment (ignacioct) [inline docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb]: Same here, I would write `FeedbackDataset`. As we are writing it as code, it should match our class name

File: docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb [hunk 3]
```diff
@@ -0,0 +1,628 @@
...
+    "3. Provide Human Feedback to remove bias and improve the model\n",
+    "4. Train a `reward model`"
```
Comment (ignacioct) [inline docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb]: Wouldn't put it as code, but rather with capital letters.

File: docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb [hunk 4]
```diff
@@ -0,0 +1,628 @@
...
+   ]
```
Comment (ignacioct) [inline docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb]: Also, don't forget final points after sentences in the list :)

File: docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb [hunk 5]
```diff
@@ -0,0 +1,628 @@
...
+   "metadata": {},
+   "source": [
+    "## Introduction"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Language models, despite their ability to perform various natural language processing tasks, often reflect biases and ethical concerns. These biases include a range of categories such as age, gender, race, ethnicity and others ([Huang et al., 2023](https://arxiv.org/pdf/2309.14345.pdf)), and extend to issues such as misinformation, toxicity and hallucinations.\n",
```
Comment (ignacioct) [inline docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb]:
```suggestion
    "Language models, despite their ability to perform various NLP tasks, often reflect biases and ethical concerns. These biases include a range of categories such as age, gender, race, ethnicity and others ([Huang et al., 2023](https://arxiv.org/pdf/2309.14345.pdf)), and extend to issues such as misinformation, toxicity and hallucinations.\n",
```

File: docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb [hunk 6]
```diff
@@ -0,0 +1,628 @@
...
+    "\n",
+    "The root of these concerns lies in the fact that language models are trained on large datasets that replicate real-world characteristics, inadvertently perpetuating these biases and creating false associations. This leads to both technical and ethical problems, including the risk of reinforcing societal prejudices against marginalised groups.\n",
```
Comment (ignacioct) [inline docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb]: Really like this sentence 👍🏼

File: docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb [hunk 7]
```diff
@@ -0,0 +1,628 @@
...
+    "\n",
+    "Addressing these biases requires intervention that can be at different stages of model training and output generation, as suggested by several studies ([(Yeh et al., 2023)](https://aclanthology.org/2023.rocling-1.37.pdf), [(Liang et al., 2021)](https://proceedings.mlr.press/v139/liang21a/liang21a.pdf), [(Garimella et al., 2021)](https://aclanthology.org/2021.findings-acl.397.pdf)). As retraining a language model is not efficient, a notable strategy is the incorporation of human feedback. In this method, the model's responses are evaluated by human raters, and the feedback is used to develop a reward model. This reward model, in turn, guides a reinforcement learning process to adjust the parameters of the language model.\n",
```
Comment (ignacioct) [inline docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb]:
```suggestion
    "Addressing these biases requires intervention that can be at different stages of model training and output generation, as suggested by several studies ([(Yeh et al., 2023)](https://aclanthology.org/2023.rocling-1.37.pdf), [(Liang et al., 2021)](https://proceedings.mlr.press/v139/liang21a/liang21a.pdf), [(Garimella et al., 2021)](https://aclanthology.org/2021.findings-acl.397.pdf)). As retraining a language model is not efficient, a notable strategy is the incorporation of human feedback. In this method, the model's responses are evaluated by human annotators, and the feedback is used to develop a reward model. This reward model, in turn, guides a reinforcement learning process to adjust the parameters of the language model.\n",
```

Unsure if raters exists, we can maybe use annotators, as rating is annotating data at the end of the day

File: docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb [hunk 8]
```diff
@@ -0,0 +1,628 @@
...
+    "\n",
+    "The essence of this approach is to make the outputs of the language model more closely match human norms and values. In this way, bias is reduced, robustness is increased, and the model's outputs are more ethically aligned and socially responsible, minimizing their potential negative impact on society."
```
Comment (ignacioct) [inline docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb]:
```suggestion
    "The essence of this approach is to match the outputs of the language model more closely with human norms and values. In this way, bias is reduced, robustness is increased, and the model's outputs are more ethically aligned and socially responsible, minimizing their potential negative impact on society."
```

File: docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb [hunk 9]
```diff
@@ -0,0 +1,628 @@
...
+    "from langchain.llms import OpenAI\n",
+    "from langchain.chains.base import Chain\n",
+    "from langchain.vectorstores import FAISS\n",
+    "from langchain.prompts import PromptTemplate\n",
+    "from langchain.embeddings import OpenAIEmbeddings\n",
+    "from langchain.document_loaders import PyPDFLoader\n",
+    "from langchain.chains import RetrievalQA, load_chain\n",
+    "from langchain.text_splitter import RecursiveCharacterTextSplitter\n",
+    "\n",
+    "from giskard import Dataset, Model, scan"
```
Comment (ignacioct) [inline docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb]: By convention, inside each of the blocks, we follow alphabetical order for imports. Doesn't the formatted automatically, does it? If not, is just moving around some stuff :) I can push myself the change if you are unsure.

File: docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb [hunk 10]
```diff
@@ -0,0 +1,628 @@
...
+    "\n",
+    "Your answer:\n",
+    "\"\"\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now, we will create a QA system which retrieves the data from our report and uses the LLM to answer questions. For this purpose, we will use `FAISS` which storages the chuncks of context and `LangChain` which integrates the LLM with the retriever."
```
Comment (ignacioct) [inline docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb]:
```suggestion
    "Now, we will create a QA system, which retrieves the data from our report and uses the LLM to answer questions. For this purpose, we will use `FAISS`, for storing the chunks of context, and `LangChain`, to integrate the LLM with the retriever."
```

File: docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb [hunk 11]
```diff
@@ -0,0 +1,628 @@
...
+    "llm = OpenAI(model=LLM_NAME, temperature=0)\n",
+    "prompt = PromptTemplate(template=PROMPT_TEMPLATE, input_variables=[\"question\", \"context\"])\n",
+    "qa_system = RetrievalQA.from_llm(llm=llm, retriever=get_context_storage().as_retriever(), prompt=prompt)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "After creating the QA system, we will create a custom `Giskard.Model` object which will be used to test the LLM. After that, we [will wrap it up indicating the needed parameters](https://docs.giskard.ai/en/latest/open_source/scan/scan_llm/index.html): the input `model` (our `qa_system`); the `model_type`, as working with LLM is always `text_generation`; the `name` (used as metadata); the `description` of the model used to generate the testing prompt and the `feature_names` which will be the columns of our dataset."
```
Comment (ignacioct) [inline docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb]:
```suggestion
    "Once we have the QA system ready, we will create a custom `Giskard.Model` object which will be used to test the LLM. After that, we [will wrap it up indicating the needed parameters](https://docs.giskard.ai/en/latest/open_source/scan/scan_llm/index.html): the input `model` (our `qa_system`); the `model_type`, as working with LLM is always `text_generation`; the `name` (used as metadata); the `description` of the model used to generate the testing prompt and the `feature_names` which will be the columns of our dataset."
```

File: docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb [hunk 12]
```diff
@@ -0,0 +1,628 @@
...
+    "            \"question_name\": \"new-instruction\",\n",
+    "            \"value\": row['input_question'],\n",
+    "        },\n",
+    "        {\n",
+    "            \"question_name\": \"new-response\",\n",
+    "            \"value\": row['model_output'],\n",
+    "        }\n",
+    "    ],\n",
+    "    )\n",
+    "    for index, row in df.iterrows()\n",
```
Comment (ignacioct) [inline docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb]:
```suggestion
    "    for _, row in df.iterrows()\n",
```

File: docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb [hunk 13]
```diff
@@ -0,0 +1,628 @@
...
+   "source": [
+    "# Create and push a feedback dataset\n",
+    "dataset = rg.FeedbackDataset(\n",
+    "    fields=[rg.TextField(name=\"instruction\"), rg.TextField(name=\"response\")],\n",
+    "    questions=[\n",
+    "        rg.TextQuestion(name=\"new-instruction\", title=\"Write a helpful, harmless, accurate instruction for the user response\"),\n",
+    "        rg.TextQuestion(name=\"new-response\", title=\"Write a helpful, harmless, accurate response to the user question\"),\n",
+    "    ],\n",
+    ")\n",
+    "dataset = dataset.push_to_argilla(name=\"bias_dataset\", workspace=\"argilla\")"
```
Comment (ignacioct) [inline docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb]: If we push some data to argilla for annotation, we have to pull from it afterwards. If not, we are not using the annotated version.

Comment (ignacioct) [inline docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb]: Maybe, also find more meaningful names for the datasets and dataframes than dataset and ds. It makes it easier to differentiate the variables without having additional context.

File: docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb [hunk 14]
```diff
@@ -0,0 +1,628 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from argilla.feedback import ArgillaTrainer\n",
+    "\n",
+    "trainer = ArgillaTrainer(\n",
+    "    dataset=ds,\n",
+    "    task=task,\n",
+    "    framework=\"trl\",\n",
+    "    model=\"distilroberta-base\",\n",
```
Comment (ignacioct) [inline docs/_source/tutorials_and_integrations/tutorials/feedback/monitoring-bias-ethics-rlhf.ipynb]: Same here, more meaningful names for ds and task maybe

**Review by ignacioct — CHANGES_REQUESTED** [review]
Looking nice! I've made a few comments on the writing and some code, I still have to test the code. Some minor general points:

- Maybe write a conclusion afterwards, where you can wrap up the job you've done, what the user must take away, and some links to related Argilla content. 
- Some more meaningful naming for some variables, for example I'm unsure of the difference between dataset and ds, apart from the fact that they appear in different points of the tutorial. 
- Maybe generate some cell outputs so users can see how the code works.

Apart from that, maybe give it a grammar correction. I really like the writing stile, very concise and to the point.

**Review by davidberenstein1957 — COMMENTED** [review]
@sdiazlor, before closing I'll also check this with "mps" on mac. We might have some m1 issues though 🥲

**Comment by sdiazlor** [comment]
> @sdiazlor, before closing I'll also check this with "mps" on mac. We might have some m1 issues though 🥲

Perfect, that would be really useful!
